control-base/src/main/java/fi/nls/oskari/util/WFSGetLayerFields.java

```diff
@@ -0,0 +1,104 @@
+package fi.nls.oskari.util;
+
+import fi.nls.oskari.domain.map.OskariLayer;
+import fi.nls.oskari.service.ServiceException;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+public class WFSGetLayerFields {
+
+    /**
+     * Return fields information for the WFS layer
+     *
+     * The result is constructed as:
+     * {
+     *     "attributes": [{
+     *         "name": "attribute-name-1",
+     *         "type": "attribute-type-1"
+     *     },{
+     *         "name": "attribute-name-2",
+     *         "type": "attribute-type-2"
+     *     }, ...],
+     *     "geometryField": "geometry"
+     * }
+     *
+     * @param layer
+     * @return
+     * @throws ServiceException
+     */
+    public static JSONObject getLayerFields(OskariLayer layer) throws ServiceException {
+        if (layer.getVersion().startsWith("3")) {
+            return getCollectionFields(layer);
+        } else {
+            return getFeatureTypeFields(layer);
+        }
+    }
+
+    /**
+     * Return fields information for WFS 3.x layer
+     * @param layer
+     * @return
+     * @throws ServiceException
+     */
+    private static JSONObject getCollectionFields(OskariLayer layer) throws ServiceException {
+        final String url = layer.getUrl() + "collections/" + layer.getName() + "/queryables";
```

Comment:
The queryables endpoint is used for describing for example attributes that can be used to search items. It isn't a comprehensive list of attributes that the items in the collection have. The best we can do here instead is querying the /items endpoint for the first 10 or so features and gather the set of attribute names detected from those features. That should be reasonably accurate.

Reply:
Makes sense